### ISSUE TITLE
Async methods with implementation in local functions

### ISSUE THREAD

**Issue by atifaziz** [zompinc/sync-method-generator#1]
First of all, this is a really cool idea and I'm impressed with the results on my initial exploration. I'm strongly considering using this for [MoreLINQ](https://github.com/morelinq/MoreLINQ) so one can just write the async version of operators and the sync versions would get generated, but there's a problem. Generally, public async (iterator) methods are split into two parts in order to do eager argument validation; that is, a public method that does argument validation and a private local function that contains the actual core functionality. For example, the “[Dissecting the local functions in C# 7](https://devblogs.microsoft.com/premier-developer/dissecting-the-local-functions-in-c-7/)” article has the following example at the end of the “[Use Case 1: eager preconditions in iterator blocks](https://devblogs.microsoft.com/premier-developer/dissecting-the-local-functions-in-c-7/#use-case-1-eager-preconditions-in-iterator-blocks)” section (granted it's not async, but “[Use Case 2: eager preconditions in async methods](https://devblogs.microsoft.com/premier-developer/dissecting-the-local-functions-in-c-7/#use-case-2-eager-preconditions-in-async-methods)” calls that one out too): 

```c#
public static IEnumerable<string> ReadLineByLine(string fileName)
{
    if (string.IsNullOrEmpty(fileName)) throw new ArgumentNullException(nameof(fileName));

    return ReadLineByLineImpl();

    IEnumerable<string> ReadLineByLineImpl()
    {
        foreach (var line in File.ReadAllLines(fileName))
        {
            yield return line;
        }
    }
}
```

Since C# 9, it's possible to [apply attributes to location functions](https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/proposals/csharp-9.0/local-function-attributes). I was hoping one could just apply `[CreateSyncVersion]` to a local function (that's either an async method or an async iterator method), but that doesn't work. When I looked through the source, I didn't find anything predicated on `LocalFunctionStatement` so that explains that, but I was wondering if this is something you're planning to address?

**Comment by virzak** [comment]
@atifaziz thanks for noticing the source generator and filing the very first issue. I'm sure a lot of very obvious scenarios are missing at this point and thanks for pointing this one out. It would be great to have MoreLINQ utilize this library.

2b7d000 (already on NuGet) handles both use case 1 and 2 without applying the attribute to the local method - applying it to the outer function is the only thing you need to do.

If there's another use case where the attribute should only be applied to the local function, please let me know.

**Comment by atifaziz** [comment]
> [2b7d000](https://github.com/zompinc/sync-method-generator/commit/2b7d000be4032d2932fed9c86927fdd31ffe711c) (already on NuGet) handles both use case 1 and 2 without applying the attribute to the local method - applying it to the outer function is the only thing you need to do.

Hey, thanks for your super quick follow-up on this. I tried out the [new version](https://www.nuget.org/packages/Zomp.SyncMethodGenerator/0.1.13-alpha) in MoreLINQ again and it doesn't seem to be working for async iterators.

I checked the tests added in 2b7d000be4032d2932fed9c86927fdd31ffe711c and something seems fishy in there. For example, the following command:

    git ls-tree -r --name-only 2b7d000 | grep -F EagerPrecondition

gives 3 snapshot files:

    tests/Zomp.SyncMethodGenerator.Tests/Snapshots/IntegrationTesting.EagerPreconditionInIteratorBlock#CreateSyncVersionAttribute.g.verified.cs
    tests/Zomp.SyncMethodGenerator.Tests/Snapshots/IntegrationTesting.EagerPreconditionsInAsyncMethod#CreateSyncVersionAttribute.g.verified.cs
    tests/Zomp.SyncMethodGenerator.Tests/Snapshots/IntegrationTesting.EagerPreconditionsInAsyncMethod#Test.Stuff.GetAllTextAsync.g.verified.cs  

While there's a pair for the `EagerPreconditionsInAsyncMethod` test, there's just one (attribute) file for the `EagerPreconditionInIteratorBlock` test.

I also grep'd for `ReadLineByLineAsync` which is the iterator in the `EagerPreconditionInIteratorBlock` test case:

    git grep -F ReadLineByLineAsync 2b7d000

and got just one hit that's in the original test file `IntegrationTesting.cs`.

Am I missing or doing something wrong? I did apply the attribute to the outer method.

**Comment by virzak** [comment]
Thanks for the catch, looks like the test case was broken. With the latest 6074f0b commit you can see that the [verified generated code](https://github.com/zompinc/sync-method-generator/blob/6074f0be8b6fc42a44af02d8908f2a1461a2c4df/tests/Zomp.SyncMethodGenerator.Tests/Snapshots/IntegrationTesting.EagerPreconditionInIteratorBlock%23Test.Stuff.ReadLineByLineAsync.g.verified.cs#L7-L28) is legit.

The `src` directory hasn't been modified, so this was just a problem with the test.

This fix doesn't explain why you weren't able to generate your sync method though, so perhaps you could either paste your async iterator scenario here, or just point me to the relevant lines in MoreLinq library.

**Comment by atifaziz** [comment]
So sadly I was relying too much on the generated code node showing up in the **Solution Explorer**:

<img width="338" alt="image" src="https://user-images.githubusercontent.com/20511/201754227-631107e4-e863-4351-9e82-003ed633c19c.png">

It was working before and I could see the generated code appear there with the earlier version, but not since updating to the latest version of the package. I've seen this happen before so it's definitely a Visual Studio hiccup. I checked the output assembly and it does indeed have the synchronous version generated of the iterator! I should have done that before sounding the alarm. On a positive note, we got the test case fixed. 😉

I'll close this now since the reported problem is solved. Thanks again for the super quick turnarounds!

---
PS After restarting Visual Studio, the generate code node for synchronous version of the iterator is showing up once more.
